### DISCUSSION TITLE
Feature Request: Documentation should clarify Python read access thread-safety

### DISCUSSION THREAD

**Discussion by captaincoordinates** [duckdb/duckdb#12817] [Ideas]
Python client API [documentation](https://duckdb.org/docs/api/python/overview#using-connections-in-parallel-python-programs) states that the `DuckDBPyConnection` object is not thread-safe and therefore a cursor should be created for parallel _writes_.

Documentation should clarify expected _read_ behaviour. For example, if multiple threads are using the same `DuckDBPyConnection` object to read data does this also need special handling?
- Can multiple parallel callers receive the wrong result set?
- Is this affected by whether the data source is a DuckDB file, in-memory database, or Parquet or other format source data?

I'm happy to create a PR for documentation updates if necessary, but first I would need someone to tell me how DuckDB behaves in this scenario. Unfortunately I don't know enough C++ to check for myself.

**Comment by Tishj** [comment]
Any parallel query execution on a single connection will break.

Only one query execution can be active at a time on a given connection.
It does not matter whether these are read or write queries

**Reply by captaincoordinates** [reply]
@Tishj thanks for the clarification, and does the same advice apply in both scenarios, i.e. that a cursor will permit parallel query execution? Could this change based on the data source?

**Reply by Tishj** [reply]
Yes it applies in both scenarios, read and write should be save as long as separate cursors are used

**Comment by jgrg** [comment]
I've just been writing some code where I was `yield`ing pairs of rows from a query in one function to be processed by another function which then inserts a row into another table (which is not part of the query). I created a separate cursor for the second function, but the Python script would lock up after < 1000 rows. It sounds like this should have worked?

I changed to doing a `fetchall()` on the query and `yield`ing pairs of rows from the array, but it was disappointing not to be able to use the cursor to stream rows.

**Reply by Tishj** [reply]
Do you have a reproduction you can share for this?
I can't really discern whether this is intended or not from just this description

**Reply by jgrg** [reply]
I'll try to make a reproducible test…

**Comment by foldager** [comment]
I was facing the same question: trying to figure out the thread-safety for concurrent reads in python.

The documentation is not easy to navigate on this. I finally found the following in the [DB API docs](https://duckdb.org/docs/api/python/dbapi.html)

> The standard DuckDB Python API provides a SQL interface compliant with the [DB-API 2.0 specification described by PEP 249](https://www.python.org/dev/peps/pep-0249/) …

I checked at runtime that `duckdb.threadsafety=1`, so according to the [python database API 2.0 threadsafety](https://peps.python.org/pep-0249/#threadsafety) spec

> Threads may share the module, but not connections.

However, other parts of the docs do not agree with that. 

- Connections (but not cursors) are shared between threads in the guide [Multiple Python Threads](https://duckdb.org/docs/guides/python/multiple_threads.html)
- The page on [connections and concurrency](https://duckdb.org/docs/connect/concurrency) says that a single process can concurrently read/write to a db, that is, a connection can be shared across threads.

Which makes me question if those pages should be updated, or if `duckdb.threadsafty` should be updated to `2` to indicate that:
> Threads may share the module and connections.

Also, the [documentation](https://duckdb.org/docs/api/python/reference/#duckdb.threadsafety) for `duckdb.threadsafety` indicates it is a bool, which threw me off at first. It would be nice if it could document the value of `duckdb.threadsafety` or a least what the possible values mean.

**Reply by Tishj** [reply]
`Threads may share the module, but not connections` is entirely correct.
`cursor()` is kind of an exception to this, but perhaps it shouldn't be.

It's made thread safe mainly because it was already being used that way

**Comment by hussainsultan** [comment]
Here is a reproducible example that shows inconsistent behavior when sharing a connection across threads. 

```python
import duckdb
import datafusion
import threading
import time
from queue import Queue

def test_duckdb_connection_thread_safety():
    print("\nTesting DuckDB Connection Thread Safety")
    print("=====================================")
    
    # shared connection
    conn = duckdb.connect(':memory:')
    results = Queue()
    
    def worker(id: int):
        try:
            result = conn.execute(f"""
                SELECT {id} as thread_id, 
                       range as val 
                FROM range(5)
            """).fetchall()
            results.put(("success", id, result))
        except Exception as e:
            results.put(("error", id, str(e)))
    
    threads = []
    for i in range(3):
        t = threading.Thread(target=worker, args=(i,))
        threads.append(t)
        t.start()
    
    for t in threads:
        t.join()
        
    conn.close()
    
    print("\nDuckDB Results:")
    while not results.empty():
        status, thread_id, result = results.get()
        print(f"Thread {thread_id}: {status} - {result}")

if __name__ == "__main__":
    print("Thread Safety: DuckDB")
    test_duckdb_connection_thread_safety()
```
Sometimes the results are inconsistent:
```
DuckDB Results:
Thread 0: success - []
Thread 2: success - []
Thread 1: success - [(1, 0), (1, 1), (1, 2), (1, 3), (1, 4)]
```
and worse, sometimes it will fail with a `Segmentation fault (core dumped)`

I guess my expectation would be that it should not be possible to share the threads. For example, in DataFusion if I try to share the `SessionContext` it results in an Already Borrowed error. Is that correct?

**Reply by Tishj** [reply]
Thats expected, `cursor()` exists for this, you can't share connections across threads

`cursor()` just creates a new connection so each thread will have its own

**Reply by hussainsultan** [reply]
Gotcha and thank you!
